drivers/spi/spi_context.h

```diff
@@ -113,78 +128,140 @@ static inline void spi_context_complete(struct spi_context *ctx, int status)
 			k_poll_signal(ctx->signal, status);
 		}
 
-		if (!(ctx->config->operation & SPI_LOCK_ON)) {
+		if (!(ctx->config->operation & (SPI_LOCK_ON |
+						SPI_OP_MODE_SLAVE))) {
 			k_sem_give(&ctx->lock);
 		}
 	}
 #else
 	ctx->sync_status = status;
 	k_sem_give(&ctx->sync);
-#endif
+#endif /* CONFIG_SPI_ASYNC */
+}
+
+static inline bool spi_context_is_cs_controlled(struct spi_context *ctx)
+{
+	return (ctx->config->cs && ctx->config->cs->gpio_dev);
+}
+
+static inline int spi_context_cs_active_value(struct spi_context *ctx)
+{
+	if (ctx->config->operation & SPI_CS_ACTIVE_HIGH) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static inline int spi_context_cs_inactive_value(struct spi_context *ctx)
+{
+	if (ctx->config->operation & SPI_CS_ACTIVE_HIGH) {
+		return 0;
+	}
+
+	return 1;
 }
 
 static inline void spi_context_cs_configure(struct spi_context *ctx)
 {
-	if (ctx->config->cs && ctx->config->cs->gpio_dev) {
+	if (!spi_context_is_cs_controlled(ctx)) {
+		return;
+	}
+
+#ifdef CONFIG_SPI_SLAVE
+	if (spi_context_is_slave(ctx)) {
 		gpio_pin_configure(ctx->config->cs->gpio_dev,
-				   ctx->config->cs->gpio_pin, GPIO_DIR_OUT);
-		gpio_pin_write(ctx->config->cs->gpio_dev,
-			       ctx->config->cs->gpio_pin, 1);
-	} else {
-		SYS_LOG_INF("CS control inhibited (no GPIO device)");
+				   ctx->config->cs->gpio_pin,
+				   GPIO_DIR_IN | GPIO_INT |
+				   GPIO_INT_DOUBLE_EDGE | GPIO_INT_DEBOUNCE);
+		return;
 	}
+#endif /* CONFIG_SPI_SLAVE */
+
+	gpio_pin_configure(ctx->config->cs->gpio_dev,
+			   ctx->config->cs->gpio_pin,
+			   GPIO_DIR_OUT |
+			   (spi_context_cs_inactive_value(ctx) ?
+			    GPIO_PUD_PULL_UP : GPIO_PUD_PULL_DOWN));
+	gpio_pin_write(ctx->config->cs->gpio_dev,
+		       ctx->config->cs->gpio_pin,
+		       spi_context_cs_inactive_value(ctx));
 }
 
 static inline void spi_context_cs_control(struct spi_context *ctx, bool on)
 {
-	if (ctx->config->cs && ctx->config->cs->gpio_dev) {
+	if (!spi_context_is_cs_controlled(ctx)) {
+		return;
+	}
+
+#ifdef CONFIG_SPI_SLAVE
+	if (spi_context_is_slave(ctx)) {
 		if (on) {
-			gpio_pin_write(ctx->config->cs->gpio_dev,
-				       ctx->config->cs->gpio_pin, 0);
-			k_busy_wait(ctx->config->cs->delay);
+			gpio_add_callback(ctx->config->cs->gpio_dev,
+					  &ctx->cs_cb);
+			gpio_pin_enable_callback(ctx->config->cs->gpio_dev,
+						 ctx->config->cs->gpio_pin);
 		} else {
-			if (ctx->config->operation & SPI_HOLD_ON_CS) {
-				return;
-			}
+			gpio_pin_disable_callback(ctx->config->cs->gpio_dev,
+						  ctx->config->cs->gpio_pin);
+			gpio_remove_callback(ctx->config->cs->gpio_dev,
+					     &ctx->cs_cb);
+		}
+	}
```

Comment:
Missing `return` at the end of this block, I guess.

Reply:
Indeed